compiler/noirc_frontend/src/tests/visibility.rs

```diff
@@ -146,3 +146,110 @@ fn does_not_error_if_calling_private_struct_function_from_same_module() {
     "#;
     assert_no_errors(src);
 }
+
+#[test]
+fn error_when_accessing_private_struct_field() {
+    let src = r#"
+    mod moo {
+        pub struct Foo {
+            x: Field
+        }
+    }
+
+    fn foo(foo: moo::Foo) -> Field {
+        foo.x
+    }
+
+    fn main() {}
+    "#;
+
+    let errors = get_program_errors(src);
+    assert_eq!(errors.len(), 1);
+
+    let CompilationError::ResolverError(ResolverError::PathResolutionError(
+        PathResolutionError::Private(ident),
+    )) = &errors[0].0
+    else {
+        panic!("Expected a private error");
+    };
+
+    assert_eq!(ident.to_string(), "x");
+}
+
+#[test]
+fn does_not_error_when_accessing_private_struct_field_from_nested_module() {
+    let src = r#"
+    struct Foo {
+        x: Field
+    }
+
+    mod nested {
+        fn foo(foo: super::Foo) -> Field {
+            foo.x
+        }
+    }
+
+    fn main() {
+        let _ = Foo { x: 1 };
+    }
+    "#;
+    assert_no_errors(src);
+}
+
+#[test]
+fn error_when_using_private_struct_field_in_constructor() {
+    let src = r#"
+    mod moo {
+        pub struct Foo {
+            x: Field
+        }
+    }
+
+    fn main() {
+        let _ = moo::Foo { x: 1 };
+    }
+    "#;
+
+    let errors = get_program_errors(src);
+    assert_eq!(errors.len(), 1);
+
+    let CompilationError::ResolverError(ResolverError::PathResolutionError(
+        PathResolutionError::Private(ident),
+    )) = &errors[0].0
+    else {
+        panic!("Expected a private error");
+    };
+
+    assert_eq!(ident.to_string(), "x");
+}
+
+#[test]
+fn error_when_using_private_struct_field_in_struct_pattern() {
```

Comment:
Please add a test for the case(s) where the field is `pub x`

Reply:
Sounds good. That said, it's kind of tested already in the standard library because if that doesn't work the stdlib produces warnings. But I'll add a test 👍